src/main/java/io/cryostat/credentials/Credentials.java

```diff
@@ -39,19 +44,62 @@
 import jakarta.ws.rs.Path;
 import jakarta.ws.rs.core.Context;
 import jakarta.ws.rs.core.UriInfo;
+
+import org.eclipse.microprofile.config.inject.ConfigProperty;
 import org.jboss.logging.Logger;
 import org.jboss.resteasy.reactive.RestForm;
 import org.jboss.resteasy.reactive.RestPath;
 import org.jboss.resteasy.reactive.RestResponse;
 import org.jboss.resteasy.reactive.RestResponse.ResponseBuilder;
 import org.projectnessie.cel.tools.ScriptException;
 
+import com.fasterxml.jackson.databind.ObjectMapper;
+
 @Path("/api/v4/credentials")
 public class Credentials {
 
+    @ConfigProperty(name = ConfigProperties.CREDENTIALS_DIR)
+    java.nio.file.Path dir;
+
     @Inject TargetConnectionManager connectionManager;
     @Inject TargetMatcher targetMatcher;
     @Inject Logger logger;
+    @Inject ObjectMapper mapper;
+
+    @Transactional
+    void onStart(@Observes StartupEvent evt) {
+                if (!checkDir()) {
+            return;
+        }
+        try {
+            Files.walk(dir)
+                    .filter(Files::isRegularFile)
+                    .filter(Files::isReadable)
+                    .forEach(this::createFromFile);
+        } catch (IOException e) {
+            logger.error(e);
+        }
+    }
+
+    private boolean checkDir() {
+        return Files.exists(dir)
+                && Files.isReadable(dir)
+                && Files.isExecutable(dir)
+                && Files.isDirectory(dir);
+    }
+
+    private void createFromFile(java.nio.file.Path path) {
+        try (var is = new BufferedInputStream(Files.newInputStream(path))){
+            var credential = mapper.readValue(is, Credential.class);
+            var exists = Credential.find("id", credential.id).count() != 0;
```

Comment:
I'm not sure about checking for the IDs like this. The database will assign the IDs when the credentials are created, so I think this scheme can likely lead to duplicates getting created. If I manually create a Credential via API and it gets ID 1, then I shut the system down and put in a different declarative Credential definition with ID 1 into the directory, this will incorrectly reject the new declarative Credential even though it is actually new and different.

I think the right thing to do here is parse the declaratively defined Credentials, then check the database if any Credential exists with the exact same `<username, password, matchExpression>` - this tuple is what really defines the Credential.